first pass add-ref (ie pull from registered drs object)

quinnwai opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
See #15


## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of Changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

bwalsh commented:
### init failing

documentation/instruction out of date?

```
$ git-drs init --apiendpoint https://calypr-dev.ohsu.edu/ --profile calypr-dev --cred ~/.gen3/credentials-calypr-dev.json --project cbds-git_drs_test --bucket cbds
Error: unknown flag: --apiendpoint
2025/07/30 16:41:05 Error: unknown flag: --apiendpoint
$ which git-drs
/Users/walsbr/calypr/git-drs/git-drs
$ ls -l /Users/walsbr/calypr/git-drs/git-drs
-rwxr-xr-x  1 walsbr  OHSUM01\Domain Users  11108626 Jul 30 16:23 /Users/walsbr/calypr/git-drs/git-drs
$ git-drs version
git commit:
git branch:
git upstream:
build date:
version:
```

 git-drs init `--url` https://calypr-dev.ohsu.edu/ --profile calypr-dev --cred ~/.gen3/credentials-calypr-dev.json --project cbds-git_drs_test --bucket cbds
 
 works

bwalsh commented:
## confusing git status after `git drs add-ref`

```
RNB10644:git-drs-test-repo walsbr$ git drs init --anvil --terraProject terra-5566e7ca
Current server set to anvil
Git DRS initialized successfully!
To stage any configuration changes, use 'git add .drs/config.yaml'
RNB10644:git-drs-test-repo walsbr$ git drs add-ref drs://drs.anv0:v2_e68887be-c583-375a-a773-48771192c8fa
Adding reference to DRS object drs://drs.anv0:v2_e68887be-c583-375a-a773-48771192c8fa
Successfully added reference to DRS object drs://drs.anv0:v2_e68887be-c583-375a-a773-48771192c8fa as file data_tables_population_descriptor.tsv
RNB10644:git-drs-test-repo walsbr$ git status
On branch main
Your branch is up to date with 'origin/main'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   .drs/config.yaml
	modified:   boo.txt
```

* unexpected `[boo.txt](modified:   boo.txt)` - I didn't touch that file?
* unexpected - I had expected `data_tables_population_descriptor.tsv` to be in `Changes not staged for commit`

bwalsh commented:
## documentation / testing instructions

perhaps `# swap over to anvil` should be a new or different repo?

bwalsh commented:
## files created by init are not added ?

```
  (use "git add <file>..." to include in what will be committed)
	.drs/
	.gitignore
```

bwalsh commented:
## drs_downloader log placed in user's directory?
```
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	drs_downloader.log
```

bwalsh commented:
## anvil log tracked by git?

```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   .drs/anvil.log

```

quinnwai commented:
@bwalsh 

> Trouble running the release

fixed now, see the update [README](https://github.com/calypr/git-drs/pull/20/commits/abbf626cdd1bccb35dc4f175ca036d19e828d20e#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5) for more info.

> Add/Create CONTRIBUTING.md
> $ git submodule update --remote
> $ go build
that's fixed now, we're importing directly since #31 

> confusing git status after git drs add-ref
> - unexpected [boo.txt](modified:   boo.txt) - I didn't touch that file?

I was able to reproduce. Deleted the file, so that should be updated on the git-drs-test-repo repo. I think the pointer I added prolly didn't have a newline so when it localized the changes assumed it was modified.

> confusing git status after git drs add-ref
> - unexpected - I had expected data_tables_population_descriptor.tsv to be in Changes not staged for commit

That's because the population descriptor has already been committed, would you like me to make a change so that it determines whether a file has been localized?

> takes 2.5 secs to add a reference?

Yeah most of that is waiting on the API I think, you can see that reflected in the git drs log

> drs_downloader log placed in user's directory?

yeah, that's a product of calling drs_downloader as a CLI, I don't think I can change this without reimplementing what DRS downloader does.

> anvil log tracked by git?

hmm not sure why it's showing up like that, anvil.log doesn't get automatically added. Did you commit it in one of your earlier commits (eg a `git add .drs/`?)

matthewpeterkort commented:
Might have missed it but you're going to need some way of refreshing the fence token automatically if it has expired. Gen3-client does it by calling this RequestNewAccessToken function. Can't expect the user to run ```git-drs init``` every time they have a fixable issue Ex:
```
	import token "github.com/bmeg/grip-graphql/middleware"

	expiration, err := token.GetExpiration(cl.Cred.AccessToken)
	if err != nil {
		return &Resp{nil, err}
	}
	// Update AccessToken if token is old
	if expiration.Before(time.Now()) {
		r := jwt.Request{}
		r.RequestNewAccessToken(cl.Base.String()+commonUtils.FenceAccessTokenEndpoint, &cl.Cred)
	}
	if cl.Cred.AccessToken == "" {
		return &Resp{nil, fmt.Errorf("access token not found in profile config")}
	}
```

matthewpeterkort commented:
This is a bad error message:
```Error: [ERROR] unable to configure your gen3 profile: Error occurred when validating profile config: The provided apiendpoint 'https://calypr-dev.ohsu.edu' is possibly not a valid Gen3 data commons``` 

For what you get when you're not on OHSU-Secure and you attempt to connect to an internal instance.

matthewpeterkort commented:
There shouldn't be a "type" keyword under auth. The server type is gen3 as stated as the key name under servers. I shouldn't be needed to be specified twice. 

The bucket key is a bit vague and shouldn't be needed for gen3 instances. The bucket can be parsed from a fence get buckets request since one program maps to one bucket and we already know the project_id.

I would consider getting rid of the auth: key all together. Simpler is better. You can still have the "auth" structure but make it an embedded structure inside both the gen3 and the anvil structs.

```
current_server: gen3
servers:
    gen3:
        endpoint: https://calypr-dev.ohsu.edu
        auth:
            type: gen3
            profile: cbds-dev
            project_id: cbds-testmultimodal
            bucket: cbds
```

matthewpeterkort commented on utils/anvil.go:
```diff
@@ -0,0 +1,38 @@
+	"fmt"
+	"path/filepath"
+)
+
```
instead of lodging this in a utils package put it in the anvil client package so like ./client/anvil/anvil.go

quinnwai replied:
done

matthewpeterkort commented on cmd/addref/add-ref.go:
```diff
@@ -0,0 +1,230 @@
+		return nil, fmt.Errorf("Anvil server endpoint is empty in config file")
+	}
+
+	req, err := http.NewRequest("POST", cfg.Servers.Anvil.Endpoint, bytes.NewBuffer(bodyBytes))
```
regarding the 2.5s to add a remote, It is generally worth seeing if most of the time is spent in the POST request or if there is something else that is taking all the time

matthewpeterkort replied:
Ok great that checks out

bwalsh commented:
## confusing test instructions 🤔 

> git status   # make sure drs id has been added

What is the expected status?

My status

```
$ git status
On branch main
Your branch is up to date with 'origin/main'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   .gitattributes
	new file:   data_tables_sample_set.tsv

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   .drs/config.yaml

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	.drs/anvil.log
	.drs/git-drs.log
	.gitignore
```

bwalsh commented:
Potential re-usable testing script.

```
#!/bin/bash
set -e  # Exit on error
set -x  # Print each command as it's executed

# Set repo name and remote URL
REPO_NAME="git-drs-test-bw"
GIT_USER="walsbr"
# REMOTE_URL="https://source.ohsu.edu/$GIT_USER/$REPO_NAME"
REMOTE_URL="git@source.ohsu.edu:$GIT_USER/$REPO_NAME.git"

# Clean up if rerunning (don't fail if not removable)
rm -rf "$REPO_NAME" || true

# Create directory and initialize git
mkdir "$REPO_NAME"
cd "$REPO_NAME"
git init

# Initialize Git LFS and track *.foo files
git-drs init --url https://calypr-dev.ohsu.edu/ --profile calypr-dev --cred ~/.gen3/credentials-calypr-dev.json --project cbds-git_drs_test --bucket cbds
git lfs track "*.foo"

# set branch
git branch -M main

# Add remote
git remote  add  origin "$REMOTE_URL"


# Ensure .gitattributes is tracked
git add .gitattributes

# Create a test file
echo "sample content" > test.foo
echo "sample content 2" > test2.foo

# Add, commit, and push
git add *.foo
git commit -m "Initial commit with LFS-tracked .foo file"
git push -f -u origin main


```

quinnwai commented on cmd/addref/add-ref.go:
```diff
@@ -0,0 +1,230 @@
+	}
+
+	req, err := http.NewRequest("POST", cfg.Servers.Anvil.Endpoint, bytes.NewBuffer(bodyBytes))
+	if err != nil {
```
with your changes I think we're good on this

quinnwai commented:
> no version command?

Yeah I removed it

> misleading error message

Fixed now.

> downloads appear hung due to incomplete msg

steps to repro?

> confusing test instructions 🤔
> ... `new file:   data_tables_sample_set.tsv
Yes this is what should be expected, should confirm file path of the drs has been added

> no enforcement of linkage b/t gen3 project and git project

very much agree, nothing connects a project ID to a repo name. Moreover, we should be using the entire remote URL not just the repo name mint the UUIDs I've realized. Could you create an issue on this so we can track it? I think out of scope for this AnVIL work, lmk what you think.

> unexpected files
> `* transfer.log doesn't belong in user's project root dir - AFAIK transfer.log is a git-drs log`

**[WIP]** Yeah, I'm curious when the transfer.log showed up. I can investigate this. 

> `* .drs/git-drs.log should this be in .gitignore?`

sure, thoughts on making all of .drs but the .drs/config.yaml git ignored?

> upload failure
>  I believe this is caused when trying to add a file who's sha code already exists

Is this adding a file with an existing sha from the same repo or a different repo? This is mostly a known issue in my head and needs writing up. I can create an issue


**In Progress**
- [ ] **[WIP]** Yeah, I'm curious when the transfer.log showed up. I can investigate this. 
- [ ] Issue for adding a file with an existing sha from the same repo or a different repo
- [ ]

quinnwai commented on cmd/addref/add-ref.go:
```diff
@@ -0,0 +1,230 @@
+	checksums := []drs.Checksum{}
+	for k, v := range parsed.Hashes {
+		checksums = append(checksums, drs.Checksum{
+			Type:     k,
```
done, made an enum lmk if it's not idiomatic Go

quinnwai commented on cmd/transferref/main.go:
```diff
@@ -0,0 +1,227 @@
+		encoder := json.NewEncoder(os.Stdout)
+
+		for scanner.Scan() {
+			var msg map[string]interface{}
```
done

quinnwai commented on cmd/initialize/main.go:
```diff
@@ -34,54 +39,147 @@ var Cmd = &cobra.Command{
+				Anvil: &config.AnvilServer{
+					Endpoint: utils.ANVIL_ENDPOINT,
+					Auth: config.AnvilAuth{
+						Type:         config.ANVIL_TYPE,
```
updated

quinnwai commented on cmd/transferref/main.go:
```diff
@@ -0,0 +1,227 @@
+	},
+}
+
+func downloadFile(sha string) (string, error) {
```
See https://github.com/calypr/git-drs/pull/20#discussion_r2255623325 but yeah you're right. I think we can cross that bridge when we get there tho

matthewpeterkort commented:
Can we get git-dirs init command as a callable function and then have the cmd/initialize/main.go cmd call that function ? I need to be able call git-dirs init from Forge as a package function

quinnwai commented:
@bwalsh 
## Updates

- ready for your re-review
- logs are now consolidated, no transfer.log at the top level git should be written any more, made a change to [data-client](https://github.com/calypr/data-client/commit/03a71a62f561103365c8dd30e0e8ea646ce167cd) dependency to reflect this
- config is configure once: once a valid config is created, it refers to those values. For gen3, `git drs init` is now just `git drs init --creds /path/to/credentials.json` since that is required on an per-user basis every time.
- .gitignore automatically ignores all of .drs but the .drs/config.yaml and ignores drs_downloader.log
- lmk if I missed anything other than this and below

## A Couple Unanswered Threads

> > also hid git drs version since it's incomplete
> What prevents us from fixing it?

we'd have to sync it up with which I'd like to keep separate from this commit. Don't quote me but I think Liam was gonna do this along with this PR later on.

> FYI The error transferring "XXXX..." re-occurred.
> Looked at indexd logs, no error server side
> 
> The script that reproduces it:

There's a bug in that script, for some reason the `$DUPLICATED_CONTENT` is empty in `echo $DUPLICATED_CONTENT > test3.foo` so it writes an empty line instead of making a cp. If you fix that with just a `cp test.foo test 3.foo`, it will accordingly not trigger Git DRS and push just the Git pointer as expected.

Without fixing the bug, the error you run into is: file you wanna push has a existing indexd record. See issue #42 I've created for that to describe the problem. In that case, if you're ONLY pushing and are good with using that indexd record, there is a workaround, ie skipping Git DRS and pushing the git pointer only with `GIT_LFS_SKIP_PUSH=1 git push`. This is what we did to backport evotypes.

bwalsh commented:
##  Identical file workflow #36 


> There's a bug in that script, for some reason the $DUPLICATED_CONTENT is empty in echo $DUPLICATED_CONTENT > test3.foo so it writes an empty line instead of making a cp. If you fix that with just a cp test.foo test 3.foo, it will accordingly not trigger Git DRS and push just the Git pointer as expected.

Thanks.  Fixed testing script.




```
+ REPO_NAME=git-drs-test-bw
+ GIT_USER=walsbr
+ REMOTE_URL=git@source.ohsu.edu:walsbr/git-drs-test-bw.git
+ rm -rf git-drs-test-bw
+ mkdir git-drs-test-bw
+ cd git-drs-test-bw
+ git init
Initialized empty Git repository in /Users/walsbr/calypr/git-drs/git-drs-test-bw/.git/
+ git-drs init --url https://calypr-dev.ohsu.edu/ --profile calypr-dev --cred /Users/walsbr/.gen3/credentials-calypr-dev.json --project cbds-git_drs_test --bucket cbds
2025/08/11 13:05:46 Current server set to 
2025/08/11 13:05:47 .gitignore has been updated and staged
2025/08/11 13:05:47 Git DRS initialized successfully!
2025/08/11 13:05:47 To stage any configuration changes, use 'git add .drs/config.yaml'
+ git lfs track '*.foo'
Tracking "*.foo"
+ git branch -M main
+ git remote add origin git@source.ohsu.edu:walsbr/git-drs-test-bw.git
+ git add .gitattributes
++ date
+ export 'DUPLICATE_CONTENT=sample content Mon Aug 11 13:05:47 PDT 2025'
+ DUPLICATE_CONTENT='sample content Mon Aug 11 13:05:47 PDT 2025'
+ '[' -z 'sample content Mon Aug 11 13:05:47 PDT 2025' ']'
+ echo sample content Mon Aug 11 13:05:47 PDT 2025
++ date
+ echo 'sample content 2' Mon Aug 11 13:05:47 PDT 2025
+ git add test.foo test2.foo
+ git commit -m 'Initial commit with LFS-tracked .foo file'
Current server: gen3
[main (root-commit) bf8e194] Initial commit with LFS-tracked .foo file
 4 files changed, 12 insertions(+)
 create mode 100644 .gitattributes
 create mode 100644 .gitignore
 create mode 100644 test.foo
 create mode 100644 test2.foo
+ git push -f -u origin main
Uploading LFS objects: 100% (2/2), 0 B | 0 B/s, done.
To source.ohsu.edu:walsbr/git-drs-test-bw.git
 + 9dc8e0e...bf8e194 main -> main (forced update)
branch 'main' set up to track 'origin/main'.
+ echo sample content Mon Aug 11 13:05:47 PDT 2025
++ date
+ echo 'updated content 2' Mon Aug 11 13:05:51 PDT 2025
+ git commit -m 'New commit with updated LFS-tracked .foo file' test.foo test2.foo
Current server: gen3
[main b33f9b6] New commit with updated LFS-tracked .foo file
 2 files changed, 4 insertions(+), 4 deletions(-)
+ git push -f -u origin main
Uploading LFS objects: 100% (2/2), 0 B | 0 B/s, done.
To source.ohsu.edu:walsbr/git-drs-test-bw.git
   bf8e194..b33f9b6  main -> main
branch 'main' set up to track 'origin/main'.
+ mkdir cloned
+ cd cloned
+ git clone git@source.ohsu.edu:walsbr/git-drs-test-bw.git
Cloning into 'git-drs-test-bw'...
+ cd git-drs-test-bw
+ git-drs init --url https://calypr-dev.ohsu.edu/ --profile calypr-dev --cred /Users/walsbr/.gen3/credentials-calypr-dev.json --project cbds-git_drs_test --bucket cbds
2025/08/11 13:05:55 Current server set to 
2025/08/11 13:05:56 .gitignore already up to date
2025/08/11 13:05:56 Git DRS initialized successfully!
2025/08/11 13:05:56 To stage any configuration changes, use 'git add .drs/config.yaml'
+ grep version test.foo
version https://git-lfs.github.com/spec/v1
+ grep version test2.foo
version https://git-lfs.github.com/spec/v1
+ git lfs pull -I '*.foo'
+ grep sample test.foo
sample content Mon Aug 11 13:05:47 PDT 2025
sample content Mon Aug 11 13:05:47 PDT 2025
+ grep sample test2.foo
sample content 2 Mon Aug 11 13:05:47 PDT 2025
+ echo Adding 'sample content Mon Aug 11 13:05:47 PDT 2025' to test3.foo
Adding sample content Mon Aug 11 13:05:47 PDT 2025 to test3.foo
+ echo sample content Mon Aug 11 13:05:47 PDT 2025
+ git add test3.foo
+ git commit -m 'add test3' test3.foo
Current server: gen3
[main e8eab27] add test3
 1 file changed, 3 insertions(+)
 create mode 100644 test3.foo
+ echo 'Pushing test3.foo to remote'
Pushing test3.foo to remote
+ git push origin main
error transferring "443f9016a3065f7fb3dac29434cbb0c0c619db202bead820282e91fb9efd96f8": [500] Error registering file: error registering indexd record: error getting indexd object for oid 443f9016a3065f7fb3dac29434cbb0c0c619db202bead820282e91fb9efd96f8: error reading DRS object for oid 443f9016a3065f7fb3dac29434cbb0c0c619db202bead820282e91fb9efd96f8: open .drs/lfs/objects/44/3f/443f9016a3065f7fb3dac29434cbb0c0c619db202bead820282e91fb9efd96f8: no such file or directory

Uploading LFS objects:   0% (0/1), 0 B | 0 B/s, done.
error: failed to push some refs to 'source.ohsu.edu:walsbr/git-drs-test-bw.git'

```

## snapshot of .drs
```
$ tree -r .drs
.drs
├── git-drs.log
└── config.yaml
```

## fixed testing script

```
#!/bin/bash
set -e  # Exit on error
set -x  # Print each command as it's executed

# Set repo name and remote URL
REPO_NAME="git-drs-test-bw"
GIT_USER="walsbr"
# REMOTE_URL="https://source.ohsu.edu/$GIT_USER/$REPO_NAME"
REMOTE_URL="git@source.ohsu.edu:$GIT_USER/$REPO_NAME.git"

# Clean up if rerunning (don't fail if not removable)
rm -rf "$REPO_NAME" || true

# Create directory and initialize git
mkdir "$REPO_NAME"
cd "$REPO_NAME"
git init

# Initialize Git LFS and track *.foo files
git-drs init --url https://calypr-dev.ohsu.edu/ --profile calypr-dev --cred ~/.gen3/credentials-calypr-dev.json --project cbds-git_drs_test --bucket cbds
git lfs track "*.foo"

# set branch
git branch -M main

# Add remote
git remote  add  origin "$REMOTE_URL"


# Ensure .gitattributes is tracked
git add .gitattributes

# Create a test file
export DUPLICATE_CONTENT="sample content $(date)"
if [ -z "$DUPLICATE_CONTENT" ]; then
  echo "DUPLICATE_CONTENT is empty!" >&2
  exit 1
fi
echo $DUPLICATE_CONTENT > test.foo
echo "sample content 2" $(date) > test2.foo

# Add, commit, and push
git add *.foo
git commit -m "Initial commit with LFS-tracked .foo file"
git push -f -u origin main

# Create a test file
echo $DUPLICATE_CONTENT >> test.foo
echo "updated content 2" $(date) >> test2.foo

git commit -m "New commit with updated LFS-tracked .foo file" *.foo
git push -f -u origin main

# test the clone
mkdir cloned 
cd cloned
git clone $REMOTE_URL
cd $REPO_NAME

# Initialize Git LFS and track *.foo files
git-drs init --url https://calypr-dev.ohsu.edu/ --profile calypr-dev --cred ~/.gen3/credentials-calypr-dev.json --project cbds-git_drs_test --bucket cbds
grep version test.foo || echo "expected lfs pointer"
grep version test2.foo || echo "expected lfs pointer"

git lfs pull -I "*.foo"

grep sample test.foo || echo "expected content"
grep sample test2.foo || echo "expected content"

echo Adding "$DUPLICATE_CONTENT" to test3.foo
echo $DUPLICATE_CONTENT > test3.foo

git add test3.foo
git commit -m "add test3" test3.foo
echo "Pushing test3.foo to remote"
git push origin main

```

1 directory, 2 files
